File: src/rate_of_closure/web/src/model/workspaceSession.ts
/** Strict whole-workspace adapter for the browser's supported live state. */

import type { UnitSelections } from "../components/ImpactExplorerPanel";
import type { ClubSpec, ClubType } from "./club";
import { validateScenario, type ImpactScenario } from "./impact";
import {
  PRIMARY_VIEW_IDS,
  REQUIRED_PRIMARY_VIEW_IDS,
  type PrimaryViewId,
  type PrimaryViewState,
} from "./viewPreferences";
import {
  viewWorkspaceDocument,
  viewWorkspaceFromDocument,
  type ViewWorkspace,
} from "./viewWorkspace";
import {
  simulationWorkspaceDocument,
  simulationWorkspaceFromSession,
  type SimulationWorkspaceSnapshot,
} from "./workspaceSimulationSession";
import {
  torqueWorkspaceDocument,
  torqueWorkspaceFromSession,
  type TorqueWorkspaceSnapshot,
} from "./workspaceTorqueSession";

const WORKSPACE_SCHEMA = "rate_of_closure.workspace";
const WORKSPACE_VERSION = 2;
const SESSION_SCHEMA = "rate_of_closure.explorer_session";
const CLUB_SCHEMA = "rate_of_closure.club_configuration";
const SESSION_PAYLOAD_VERSION = 3;
const CLUB_PAYLOAD_VERSION = 1;
const STABLE_ID = /^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$/;
const UTC_TIMESTAMP =
  /^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}(?:\.[0-9]{1,6})?Z$/;
const CLUB_TYPES: readonly ClubType[] = ["Driver", "Wood", "Hybrid", "Iron", "Wedge", "Putter"];
const HEAD_STYLES = ["Auto", "Mallet", "Blade"] as const;
const CLUB_BOUNDS = {
  lengthM: [0.6, 1.3], headMassKg: [0.1, 0.5], loftDeg: [0, 70], lieDeg: [45, 80],
  moiAboutShaftKgM2: [5e-5, 2e-3], cgDepthM: [0, 0.08], cgHeightM: [0, 0.06],
} as const;
const UNIT_OPTIONS = {
  speed: ["mph", "m/s", "km/h"],
  rotation: ["deg/s", "rad/s", "rpm"],
  length: ["mm", "cm", "in"],
  distance: ["yd", "m", "ft"],
} as const;

export interface WorkspaceSessionSnapshot {
  readonly scenario: ImpactScenario;
  readonly club: ClubSpec;
  readonly units: UnitSelections;
  readonly simulation: SimulationWorkspaceSnapshot;
  readonly torque: TorqueWorkspaceSnapshot;
  readonly modules: PrimaryViewState;
  readonly viewWorkspace: ViewWorkspace;
}

export interface WorkspaceFileMetadata {
  readonly documentId: string;
  readonly title: string;
  readonly createdAtUtc: string;
  readonly modifiedAtUtc: string;
  readonly appVersion: string;
}

const isRecord = (value: unknown): value is Record<string, unknown> =>
  typeof value === "object" && value !== null && !Array.isArray(value);

function exactRecord(value: unknown, keys: readonly string[], context: string): Record<string, unknown> {
  if (!isRecord(value)) throw new TypeError(`${context} must be an object`);
  const actual = Object.keys(value);
  if (actual.length !== keys.length || keys.some((key) => !(key in value))) {
    throw new TypeError(`${context} has invalid fields`);
  }
  return value;
}

function objectRecord(value: unknown, context: string): Record<string, unknown> {
  if (!isRecord(value)) throw new TypeError(`${context} must be an object`);
  return value;
}

function finite(value: unknown, context: string): number {
  if (typeof value !== "number" || !Number.isFinite(value)) {
    throw new TypeError(`${context} must be finite`);
  }
  return value;
}

function scenarioDocument(scenario: ImpactScenario): Record<string, number> {
  validateScenario(scenario);
  return {
    clubhead_speed_mph: scenario.clubheadSpeedMph,
    omega_plane_dps: scenario.omegaPlaneDps,
    omega_shaft_dps: scenario.omegaShaftDps,
    lie_angle_deg: scenario.lieAngleDeg,
    com_to_face_mm: scenario.comToFaceMm,
    impact_offset_toe_mm: scenario.impactOffsetToeMm,
    impact_offset_high_mm: scenario.impactOffsetHighMm,
    contact_duration_us: scenario.contactDurationUs,
  };
}

function scenarioFromDocument(value: unknown): ImpactScenario {
  const data = exactRecord(value, [
    "clubhead_speed_mph", "omega_plane_dps", "omega_shaft_dps", "lie_angle_deg",
    "com_to_face_mm", "impact_offset_toe_mm", "impact_offset_high_mm", "contact_duration_us",
  ], "model_session.scenario");
  const scenario: ImpactScenario = {
    clubheadSpeedMph: finite(data.clubhead_speed_mph, "clubhead_speed_mph"),
    omegaPlaneDps: finite(data.omega_plane_dps, "omega_plane_dps"),
    omegaShaftDps: finite(data.omega_shaft_dps, "omega_shaft_dps"),
    lieAngleDeg: finite(data.lie_angle_deg, "lie_angle_deg"),
    comToFaceMm: finite(data.com_to_face_mm, "com_to_face_mm"),
    impactOffsetToeMm: finite(data.impact_offset_toe_mm, "impact_offset_toe_mm"),
    impactOffsetHighMm: finite(data.impact_offset_high_mm, "impact_offset_high_mm"),
    contactDurationUs: finite(data.contact_duration_us, "contact_duration_us"),
  };
  validateScenario(scenario);
  return scenario;
}

function clubDocument(club: ClubSpec): Record<string, unknown> {
  return {
    name: club.name,
    club_type: club.clubType,
    length_m: club.lengthM,
    head_mass_kg: club.headMassKg,
    loft_deg: club.loftDeg,
    lie_deg: club.lieDeg,
    moi_about_shaft_kg_m2: club.moiAboutShaftKgM2,
    cg_depth_m: club.cgDepthM,
    cg_height_m: club.cgHeightM,
    face_bulge_radius_m: club.faceBulgeRadiusM,
    face_roll_radius_m: club.faceRollRadiusM,
    head_style: club.headStyle ?? "Auto",
  };
}

function optionalFinite(value: unknown, context: string): number | null {
  return value === null ? null : finite(value, context);
}

function clubFromDocument(value: unknown): ClubSpec {
  const data = exactRecord(value, [
    "name", "club_type", "length_m", "head_mass_kg", "loft_deg", "lie_deg",
    "moi_about_shaft_kg_m2", "cg_depth_m", "cg_height_m", "face_bulge_radius_m",
    "face_roll_radius_m", "head_style",
  ], "club_configuration.data");
  if (typeof data.name !== "string" || data.name.trim().length === 0 ||
      !CLUB_TYPES.includes(data.club_type as ClubType) ||
      !HEAD_STYLES.includes(data.head_style as typeof HEAD_STYLES[number])) {
    throw new TypeError("club configuration identity is invalid");
  }
  const club: ClubSpec = {
    name: data.name,
    clubType: data.club_type as ClubType,
    lengthM: finite(data.length_m, "length_m"),
    headMassKg: finite(data.head_mass_kg, "head_mass_kg"),
    loftDeg: finite(data.loft_deg, "loft_deg"),
    lieDeg: finite(data.lie_deg, "lie_deg"),
    moiAboutShaftKgM2: finite(data.moi_about_shaft_kg_m2, "moi_about_shaft_kg_m2"),
    cgDepthM: finite(data.cg_depth_m, "cg_depth_m"),
    cgHeightM: finite(data.cg_height_m, "cg_height_m"),
    faceBulgeRadiusM: optionalFinite(data.face_bulge_radius_m, "face_bulge_radius_m"),
    faceRollRadiusM: optionalFinite(data.face_roll_radius_m, "face_roll_radius_m"),
    headStyle: data.head_style as typeof HEAD_STYLES[number],
  };
  for (const key of Object.keys(CLUB_BOUNDS) as (keyof typeof CLUB_BOUNDS)[]) {
    const [low, high] = CLUB_BOUNDS[key];
    if (club[key] < low || club[key] > high) throw new RangeError(`${key} is out of range`);
  }
  for (const [key, value] of [
    ["faceBulgeRadiusM", club.faceBulgeRadiusM],
    ["faceRollRadiusM", club.faceRollRadiusM],
  ] as const) {
    if (value !== null && (value < 0.1 || value > 2)) throw new RangeError(`${key} is out of range`);
  }
  return club;
}

function validatedUnits(value: unknown): UnitSelections {
  const units = exactRecord(value, Object.keys(UNIT_OPTIONS), "model_session.units");
  for (const key of Object.keys(UNIT_OPTIONS) as (keyof UnitSelections)[]) {
    if (!(UNIT_OPTIONS[key] as readonly unknown[]).includes(units[key])) {
      throw new TypeError(`unsupported ${key} unit`);
    }
  }
  return units as unknown as UnitSelections;
}

function validatedModules(layout: Record<string, unknown>): PrimaryViewState {
  const order = layout.module_order;
  const visible = layout.visible_module_ids;
  if (!Array.isArray(order) || order.length !== PRIMARY_VIEW_IDS.length ||
      new Set(order).size !== order.length || PRIMARY_VIEW_IDS.some((id) => !order.includes(id))) {
    throw new TypeError("module_order must contain every module exactly once");
  }
  if (!Array.isArray(visible) || visible.length === 0 || new Set(visible).size !== visible.length ||
      visible.some((id) => !PRIMARY_VIEW_IDS.includes(id as PrimaryViewId)) ||
      REQUIRED_PRIMARY_VIEW_IDS.some((id) => !visible.includes(id)) ||
      !visible.includes(layout.active_module_id)) {
    throw new TypeError("module visibility and active module are invalid");
  }
  return {
    version: 2,
    order: [...order] as PrimaryViewId[],
    visible: [...visible] as PrimaryViewId[],
    active: layout.active_module_id as PrimaryViewId,
  };
}

function payload(
  value: unknown,
  schema: string,
  versions: readonly number[],
  context: string,
): { readonly version: number; readonly data: Record<string, unknown> } {
  const envelope = exactRecord(value, ["schema", "schema_version", "data"], context);
  if (
    envelope.schema !== schema ||
    typeof envelope.schema_version !== "number" ||
    !versions.includes(envelope.schema_version)
  ) {
    throw new TypeError(`${context} has an unsupported schema`);
  }
  return {
    version: envelope.schema_version,
    data: objectRecord(envelope.data, `${context}.data`),
  };
}

function validateMetadata(value: unknown): void {
  const data = exactRecord(value, [
    "document_id", "title", "created_at_utc", "modified_at_utc", "app_version", "provenance",
  ], "metadata");
  for (const key of ["document_id", "title", "created_at_utc", "modified_at_utc", "app_version"] as const) {
    if (typeof data[key] !== "string" || data[key].trim().length === 0) {
      throw new TypeError(`metadata.${key} must be non-empty text`);
    }
  }
  if (!STABLE_ID.test(data.document_id as string)) {
    throw new TypeError("metadata.document_id must be a stable identifier");
  }
  const createdText = data.created_at_utc as string;
  const modifiedText = data.modified_at_utc as string;
  if (!UTC_TIMESTAMP.test(createdText) || !UTC_TIMESTAMP.test(modifiedText)) {
    throw new TypeError("workspace metadata timestamps must be strict UTC text");
  }
  const created = Date.parse(createdText);
  const modified = Date.parse(modifiedText);
  if (!Number.isFinite(created) || !Number.isFinite(modified) || modified < created) {
    throw new TypeError("workspace metadata timestamps are invalid");
  }
  if (!isRecord(data.provenance) || Object.entries(data.provenance).some(
    ([key, entry]) => key.trim().length === 0 || typeof entry !== "string",
  )) throw new TypeError("metadata.provenance must map text keys to text values");
}

/** Serialize the browser-supported whole workspace without hidden file authority. */
export function createWorkspaceDocument(
  snapshot: WorkspaceSessionSnapshot,
  metadata: WorkspaceFileMetadata,
): string {
  const document = {
    schema: WORKSPACE_SCHEMA,
    schema_version: WORKSPACE_VERSION,
    metadata: {
      document_id: metadata.documentId,
      title: metadata.title,
      created_at_utc: metadata.createdAtUtc,
      modified_at_utc: metadata.modifiedAtUtc,
      app_version: metadata.appVersion,
      provenance: { surface: "rate-of-closure-file-adapter/v1" },
    },
    model_session: {
      schema: SESSION_SCHEMA,
      schema_version: SESSION_PAYLOAD_VERSION,
      data: {
        scenario: scenarioDocument(snapshot.scenario),
        units: snapshot.units,
        simulation_setup: simulationWorkspaceDocument(snapshot.simulation, snapshot.club),
        torque_selection: torqueWorkspaceDocument(snapshot.torque),
      },
    },
    prescribed_torque_profiles: snapshot.torque.profiles.map((profile) =>
      profile.toJsonObject()),
    club_configuration: {
      schema: CLUB_SCHEMA,
      schema_version: CLUB_PAYLOAD_VERSION,
      data: clubDocument(snapshot.club),
    },
    variation_plan: null,
    layout: {
      module_order: snapshot.modules.order,
      visible_module_ids: snapshot.modules.visible,
      active_module_id: snapshot.modules.active,
      view_workspace: {
        schema: "rate_of_closure.view_workspace",
        schema_version: 1,
        data: viewWorkspaceDocument(snapshot.viewWorkspace),
      },
    },
  };
  parseWorkspaceDocument(JSON.stringify(document));
  return `${JSON.stringify(document, null, 2)}\n`;
}

/** Parse and validate the entire document before exposing any applicable state. */
export interface WorkspaceParseOptions {
  readonly legacySimulationFallback?: SimulationWorkspaceSnapshot;
  readonly legacyTorqueFallback?: TorqueWorkspaceSnapshot;
}

/** Parse a current file or deliberately migrate v1 with an explicit fallback. */
export function parseWorkspaceDocument(
  text: string,
  options: WorkspaceParseOptions = {},
): WorkspaceSessionSnapshot {
  if (typeof text !== "string" || text.trim().length === 0) {
    throw new TypeError("workspace file must contain JSON text");
  }
  const root = exactRecord(JSON.parse(text), [
    "schema", "schema_version", "metadata", "model_session",
    "prescribed_torque_profiles", "club_configuration", "variation_plan", "layout",
  ], "workspace");
  if (root.schema !== WORKSPACE_SCHEMA || root.schema_version !== WORKSPACE_VERSION) {
    throw new TypeError("unsupported workspace schema");
  }
  validateMetadata(root.metadata);
  if (root.variation_plan !== null) {
    throw new TypeError("variation plans are not supported by this adapter");
  }
  const sessionEnvelope = payload(
    root.model_session,
    SESSION_SCHEMA,
    [1, 2, SESSION_PAYLOAD_VERSION],
    "model_session",
  );
  const session = exactRecord(
    sessionEnvelope.data,
    sessionEnvelope.version === 1
      ? ["scenario", "units"]
      : sessionEnvelope.version === 2
        ? ["scenario", "units", "simulation_setup"]
        : ["scenario", "units", "simulation_setup", "torque_selection"],
    "model_session.data",
  );
  const clubEnvelope = payload(
    root.club_configuration,
    CLUB_SCHEMA,
    [CLUB_PAYLOAD_VERSION],
    "club_configuration",
  );
  const club = exactRecord(
    clubEnvelope.data,
    Object.keys(clubDocument({
      name: "x", clubType: "Driver", lengthM: 1, headMassKg: 0.2, loftDeg: 10,
      lieDeg: 56, moiAboutShaftKgM2: 0.0005, cgDepthM: 0.02, cgHeightM: 0.02,
      faceBulgeRadiusM: null, faceRollRadiusM: null,
    })),
    "club_configuration.data",
  );
  const layout = exactRecord(root.layout, [
    "module_order", "visible_module_ids", "active_module_id", "view_workspace",
  ], "layout");
  const viewEnvelope = exactRecord(
    layout.view_workspace, ["schema", "schema_version", "data"], "layout.view_workspace",
  );
  if (viewEnvelope.schema !== "rate_of_closure.view_workspace" || viewEnvelope.schema_version !== 1) {
    throw new TypeError("unsupported view workspace payload");
  }
  const parsedClub = clubFromDocument(club);
  const simulation: SimulationWorkspaceSnapshot = simulationWorkspaceFromSession({
    isLegacy: sessionEnvelope.version === 1,
    setupDocument: session.simulation_setup,
    club: parsedClub,
    legacyFallback: options.legacySimulationFallback,
  });
  const torque: TorqueWorkspaceSnapshot = torqueWorkspaceFromSession({
    isLegacy: sessionEnvelope.version < SESSION_PAYLOAD_VERSION,
    selectionDocument: session.torque_selection,
    profileDocuments: root.prescribed_torque_profiles,
    legacyFallback: options.legacyTorqueFallback,
  });
  return {
    scenario: scenarioFromDocument(session.scenario),
    club: parsedClub,
    units: validatedUnits(session.units),
    simulation,
    torque,
    modules: validatedModules(layout),
    viewWorkspace: viewWorkspaceFromDocument(viewEnvelope.data),
  };
}

File: src/rate_of_closure/application/workspace_session.py
"""Strict adapters between live explorer state and workspace documents."""

from __future__ import annotations

from collections.abc import Mapping
from dataclasses import asdict, dataclass, fields
from types import MappingProxyType

from rate_of_closure.club import ClubSpec, ClubType, HeadStyle
from rate_of_closure.model import ImpactScenario
from rate_of_closure.units import QUANTITY_UNITS
from rate_of_closure.view_workspace import (
    ViewWorkspace,
    workspace_from_document,
    workspace_to_document,
)

from ._workspace_validation import FrozenJsonValue
from .workspace_document import (
    VersionedPayload,
    WorkspaceDocument,
    WorkspaceLayout,
    WorkspaceMetadata,
)

EXPLORER_SESSION_SCHEMA = "rate_of_closure.explorer_session"
CLUB_CONFIGURATION_SCHEMA = "rate_of_closure.club_configuration"
SESSION_SCHEMA_VERSION = 1
CANONICAL_MODULE_IDS = (
    "explorer",
    "calculation",
    "simulation",
    "plots",
    "flight",
    "launch-monitor-analytics",
    "capability-optimization",
    "variation",
    "putting",
    "glossary",
)
_SCENARIO_FIELDS = frozenset(field.name for field in fields(ImpactScenario))
_CLUB_FIELDS = frozenset(
    {
        "name",
        "club_type",
        "length_m",
        "head_mass_kg",
        "loft_deg",
        "lie_deg",
        "moi_about_shaft_kg_m2",
        "cg_depth_m",
        "cg_height_m",
        "face_bulge_radius_m",
        "face_roll_radius_m",
        "head_style",
    }
)
_SESSION_FIELDS = frozenset({"scenario", "units"})


@dataclass(frozen=True)
class WorkspaceSessionMetadata:
    """Mutable-lifecycle metadata supplied by a file-session controller."""

    document_id: str
    title: str
    created_at_utc: str
    modified_at_utc: str
    app_version: str


@dataclass(frozen=True)
class ExplorerWorkspaceState:
    """The complete state currently supported by live File operations."""

    scenario: ImpactScenario
    club: ClubSpec
    units: Mapping[str, str]
    module_order: tuple[str, ...]
    visible_module_ids: tuple[str, ...]
    active_module_id: str
    view_workspace: ViewWorkspace

    def __post_init__(self) -> None:
        """Validate all values before a controller can mutate live widgets."""
        if not isinstance(self.scenario, ImpactScenario):
            raise TypeError("scenario must be an ImpactScenario")
        if not isinstance(self.club, ClubSpec):
            raise TypeError("club must be a ClubSpec")
        units = dict(self.units)
        if set(units) != set(QUANTITY_UNITS) or any(
            value not in QUANTITY_UNITS[key] for key, value in units.items()
        ):
            raise ValueError("units must select one declared unit per quantity")
        order = tuple(self.module_order)
        visible = tuple(self.visible_module_ids)
        if set(order) != set(CANONICAL_MODULE_IDS) or len(order) != len(
            CANONICAL_MODULE_IDS
        ):
            raise ValueError("module_order must contain every canonical module once")
        if (
            not visible
            or len(set(visible)) != len(visible)
            or not set(visible).issubset(order)
        ):
            raise ValueError("visible_module_ids must be a non-empty module subset")
        if "explorer" not in visible or self.active_module_id not in visible:
            raise ValueError("the explorer and active module must be visible")
        if not isinstance(self.view_workspace, ViewWorkspace):
            raise TypeError("view_workspace must be a ViewWorkspace")
        self.view_workspace.validate()
        object.__setattr__(self, "units", MappingProxyType(units))
        object.__setattr__(self, "module_order", order)
        object.__setattr__(self, "visible_module_ids", visible)


def _exact_mapping(
    value: object, fields: frozenset[str], context: str
) -> Mapping[str, object]:
    if not isinstance(value, Mapping) or set(value) != fields:
        raise TypeError(f"{context} has invalid fields")
    return value


def _scenario_data(scenario: ImpactScenario) -> dict[str, float]:
    return {key: float(value) for key, value in asdict(scenario).items()}


def _club_data(club: ClubSpec) -> dict[str, FrozenJsonValue]:
    data: dict[str, FrozenJsonValue] = asdict(club)
    data["club_type"] = club.club_type.value
    data["head_style"] = club.head_style.value
    return data


def _club_from_data(value: object) -> ClubSpec:
    data = dict(_exact_mapping(value, _CLUB_FIELDS, "club_configuration.data"))
    data["club_type"] = ClubType(data["club_type"])
    data["head_style"] = HeadStyle(data["head_style"])
    return ClubSpec(**data)


def document_from_state(
    state: ExplorerWorkspaceState, metadata: WorkspaceSessionMetadata
) -> WorkspaceDocument:
    """Compose one current strict document from fully validated live state."""
    if not isinstance(state, ExplorerWorkspaceState):
        raise TypeError("state must be an ExplorerWorkspaceState")
    if not isinstance(metadata, WorkspaceSessionMetadata):
        raise TypeError("metadata must be WorkspaceSessionMetadata")
    return WorkspaceDocument(
        metadata=WorkspaceMetadata(
            document_id=metadata.document_id,
            title=metadata.title,
            created_at_utc=metadata.created_at_utc,
            modified_at_utc=metadata.modified_at_utc,
            app_version=metadata.app_version,
            provenance={"surface": "rate-of-closure-file-adapter/v1"},
        ),
        model_session=VersionedPayload(
            EXPLORER_SESSION_SCHEMA,
            SESSION_SCHEMA_VERSION,
            {"scenario": _scenario_data(state.scenario), "units": dict(state.units)},
        ),
        prescribed_torque_profiles=(),
        club_configuration=VersionedPayload(
            CLUB_CONFIGURATION_SCHEMA,
            SESSION_SCHEMA_VERSION,
            _club_data(state.club),
        ),
        variation_plan=None,
        layout=WorkspaceLayout(
            module_order=state.module_order,
            visible_module_ids=state.visible_module_ids,
            active_module_id=state.active_module_id,
            view_workspace=VersionedPayload(
                "rate_of_closure.view_workspace",
                1,
                workspace_to_document(state.view_workspace),
            ),
        ),
    )


def state_from_document(document: WorkspaceDocument) -> ExplorerWorkspaceState:
    """Validate a supported whole document before returning applicable state."""
    if not isinstance(document, WorkspaceDocument):
        raise TypeError("document must be a WorkspaceDocument")
    if document.prescribed_torque_profiles:
        raise ValueError("prescribed torque profiles are not supported by this adapter")
    if document.variation_plan is not None:
        raise TypeError("variation plans are not supported by this adapter")
    session = document.model_session
    club = document.club_configuration
    if (session.schema, session.schema_version) != (
        EXPLORER_SESSION_SCHEMA,
        SESSION_SCHEMA_VERSION,
    ):
        raise ValueError("unsupported explorer session payload")
    if (club.schema, club.schema_version) != (
        CLUB_CONFIGURATION_SCHEMA,
        SESSION_SCHEMA_VERSION,
    ):
        raise ValueError("unsupported club configuration payload")
    session_data = _exact_mapping(session.data, _SESSION_FIELDS, "model_session.data")
    scenario_data = _exact_mapping(
        session_data["scenario"], _SCENARIO_FIELDS, "model_session.scenario"
    )
    units = session_data["units"]
    if not isinstance(units, Mapping):
        raise TypeError("model_session.units must be an object")
    view = document.layout.view_workspace
    if view is None or (view.schema, view.schema_version) != (
        "rate_of_closure.view_workspace",
        1,
    ):
        raise ValueError("workspace requires a supported compositor payload")
    return ExplorerWorkspaceState(
        scenario=ImpactScenario(**scenario_data),
        club=_club_from_data(club.data),
        units=units,
        module_order=document.layout.module_order,
        visible_module_ids=document.layout.visible_module_ids,
        active_module_id=document.layout.active_module_id,
        view_workspace=workspace_from_document(view.to_json_dict()["data"]),
    )


__all__ = [
    "CANONICAL_MODULE_IDS",
    "CLUB_CONFIGURATION_SCHEMA",
    "EXPLORER_SESSION_SCHEMA",
    "ExplorerWorkspaceState",
    "WorkspaceSessionMetadata",
    "document_from_state",
    "state_from_document",
]
